### PR TITLE
Add error handling to dashboard filters

### LINKED ISSUE #6531: Handle errors when loading data in dashboard filtering dropdowns (FE)
The dashboard filtering dropdowns are currently handling the happy path only. If loading data fails, we'll just display an empty list with no errors.

This will be even more problematic when pagination is implemented, as errors could occur on any page individually. In those cases, we should give users the option to retry.

### PR DESCRIPTION
**Pull request by acelaya**
Closes #6531

Depends on https://github.com/hypothesis/frontend-shared/pull/1685

Display an error message with a button to retry when loading items for any filtering dropdown fails.

The same kind of message is displayed regardless of the page that failed, so if the first page fails, we show the error after the "All {entity}" option, and if a subsequent page fails, we show the message after the already loaded options.

https://github.com/user-attachments/assets/e5bed471-b23a-48f3-9782-983b50e2c1c3

https://github.com/user-attachments/assets/e29675ae-408f-41ad-82dd-855db8850740

### Considerations

- The icon in the retry button is not the one I was looking for, but we don't have the typical "refresh" circle arrow, so I'll leave this one for now.
- When the first page is the one failing and we retry, the select transitions into a "disabled with ellispis" state. Ideally this should happen only the first time, not when retrying, but I think handling that is not worth the complexity it would introduce.

### Todo

- [x] Try to keep listbox open when "retry" is clicked.
- [x] Add tests

### Testing steps

Apply this diff to test the error when loading first page:

```diff
diff --git a/lms/static/scripts/frontend_apps/utils/api.ts b/lms/static/scripts/frontend_apps/utils/api.ts
index ad3479399..0efde41bb 100644
--- a/lms/static/scripts/frontend_apps/utils/api.ts
+++ b/lms/static/scripts/frontend_apps/utils/api.ts
@@ -196,6 +196,7 @@ export function useAPIFetch<T = unknown>(
     api: { authToken },
   } = useConfig(['api']);
 
+  const callNum = useRef(0);
   const fetcher: Fetcher<T> | undefined = path
     ? signal =>
         apiCall({
@@ -203,6 +204,16 @@ export function useAPIFetch<T = unknown>(
           path,
           params,
           signal,
+        }).then(res => {
+          if (path.includes('/api/dashboard/assignments')) {
+            callNum.current++;
+
+            if (callNum.current < 2) {
+              throw new Error('There was an error loading');
+            }
+          }
+
+          return res as T;
         })
     : undefined;
 
diff --git a/lms/views/dashboard/pagination.py b/lms/views/dashboard/pagination.py
index 680843cdd..396c7d893 100644
--- a/lms/views/dashboard/pagination.py
+++ b/lms/views/dashboard/pagination.py
@@ -15,7 +15,7 @@ LOG = logging.getLogger(__name__)
 T = TypeVar("T")
 
 
-MAX_ITEMS_PER_PAGE = 100
+MAX_ITEMS_PER_PAGE = 15
 """Maximum number of items to return in paginated endpoints"""


```

Apply the same diff but with `if (callNum.current === 2)` to test the error when loading a follow-up page.

In both cases, clicking "retry" should load the data properly.

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx [hunk 1]
```diff
@@ -187,30 +322,18 @@ export default function DashboardActivityFilters({
...
-    () =>
-      studentsResult.data?.map(({ display_name, ...s }) => ({
-        ...s,
-        display_name:
-          display_name ??
-          `Student name unavailable (ID: ${s.lms_id.substring(0, 5)})`,
-        has_display_name: !!display_name,
-      })),
-    [studentsResult.data],
-  );
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx:200]: This is now handled by the `StudentOption` local component, so that we can use APi fetch results consistently for the three filters.

File: lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx [hunk 2]
```diff
@@ -103,6 +136,113 @@ function LoadingOption({
...
+  retry: () => void;
+};
+
+/**
+ * Indicates an error occurred while loading filters, and presents a CTA to retry.
+ */
+function LoadingError({ entity, retry }: LoadingErrorProps) {
+  return (
+    <div
+      className="flex gap-2 items-center py-2 px-4 mb-1"
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx:150]: I see the right edge of the retry button is aligned with the right edge of the checkbox above. However because the button has a light background, the eye is drawn to the alignment of the button's text with the checkbox which is off.

Button alignment px-4

Using `pl-4` and `pr-2.5` with an `sm` button aligns the button's text with the checkbox's right edge, which I think feels better aligned.

Button alignment 2

**Review by robertknight — APPROVED** [review]
I had a query on the `tabIndex` in `LoadingOption` and a comment about the button size/alignment. Also the focus workaround could use a more detailed explanation in the code comments. Otherwise this looks good and works as expected.

File: lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx [hunk 3]
```diff
@@ -103,6 +136,113 @@ function LoadingOption({
...
+      data-testid={`${entity}-error`}
+      // Make this element "focusable" so that clicking on it does not cause
+      // the listbox containing it to be closed
+      tabIndex={-1}
+    >
+      <span className="italic text-red-error">Error loading {entity}</span>
+      <Button icon={RefreshIcon} onClick={retry} size="xs">
```
[the remark this answers is not in the document]
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx:157]: Yes. I wanted to use the usual "circular" arrow, but we don't have such icon, so I decided to leave that out of this PR.

File: lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx [hunk 4]
```diff
@@ -103,6 +136,113 @@ function LoadingOption({
...
+/**
+ * Indicates an error occurred while loading filters, and presents a CTA to retry.
+ */
+function LoadingError({ entity, retry }: LoadingErrorProps) {
+  return (
+    <div
+      className="flex gap-2 items-center py-2 px-4 mb-1"
+      data-testid={`${entity}-error`}
+      // Make this element "focusable" so that clicking on it does not cause
+      // the listbox containing it to be closed
```
[the remark this answers is not in the document]
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx:153]: This prevents the listbox to close if you click anywhere in the error message which is not the button.

TL;DR;

When clicking a non-focusable item (like a `div` without `tabIndex` explicitly set), the previously focused item looses focus, dispatching a focusout event, but the event's `relatedTarget` is null in that case, causing `useFocusAway` to close the listbox.

Basically, the workaround we used for the button and this, are solutions for two different problems which have the same effect.

These two things made it even harder for me to originally isolate the root cause because I thought I was dealing with one problem, not two.
